torch/testing/_internal/common_device_type.py

```diff
@@ -481,14 +481,19 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
     # export PYTORCH_TESTING_DEVICE_ONLY_FOR=cuda,cpu
     # export PYTORCH_TESTING_DEVICE_EXCEPT_FOR=xla
     if only_for is None:
-        only_for = os.getenv("PYTORCH_TESTING_DEVICE_ONLY_FOR", ""). split(",")
+        only_for = os.getenv("PYTORCH_TESTING_DEVICE_ONLY_FOR", None)
+        if only_for is not None:
+            only_for = only_for.split(',')
+
     if except_for is None:
-        except_for = os.getenv("PYTORCH_TESTING_DEVICE_EXCEPT_FOR", ""). split(",")
+        except_for = os.getenv("PYTORCH_TESTING_DEVICE_EXCEPT_FOR", None)
+        if except_for is not None:
+            except_for = except_for.split(',')
 
     # Creates device-specific test cases
     for base in device_type_test_bases:
         # Skips bases listed in except_for
-        if except_for and only_for:
+        if except_for is not None and only_for is not None:
```

Comment:
This strictly speaking is not needed is it?

Reply:
It is needed.
if both except_for and only_for are specified (i.e. both are not None).
Did I miss something?

Anyways updating the code to the following diff
```patch
--- a/torch/testing/_internal/common_device_type.py
+++ b/torch/testing/_internal/common_device_type.py
@@ -476,24 +476,21 @@ def instantiate_device_type_tests(generic_test_class, scope, except_for=None, on
     generic_members = set(generic_test_class.__dict__.keys()) - set(empty_class.__dict__.keys())
     generic_tests = [x for x in generic_members if x.startswith('test')]
 
+    split_if_not_empty = lambda x: x.split(",") if len(x) != 0 else []
     # Derive defaults from environment variables if available, default is still none
     # Usage:
     # export PYTORCH_TESTING_DEVICE_ONLY_FOR=cuda,cpu
     # export PYTORCH_TESTING_DEVICE_EXCEPT_FOR=xla
     if only_for is None:
-        only_for = os.getenv("PYTORCH_TESTING_DEVICE_ONLY_FOR", None)
-        if only_for is not None:
-            only_for = only_for.split(',')
+        only_for = split_if_not_none(os.getenv("PYTORCH_TESTING_DEVICE_ONLY_FOR", ''))
 
     if except_for is None:
-        except_for = os.getenv("PYTORCH_TESTING_DEVICE_EXCEPT_FOR", None)
-        if except_for is not None:
-            except_for = except_for.split(',')
+        except_for = split_if_not_none(os.getenv("PYTORCH_TESTING_DEVICE_EXCEPT_FOR", ''))
 
     # Creates device-specific test cases
     for base in device_type_test_bases:
         # Skips bases listed in except_for
-        if except_for is not None and only_for is not None:
+        if except_for != [] and only_for != []:
             assert base.device_type not in except_for or base.device_type not in only_for,\
                 "same device cannot appear in except_for and only_for"
         if except_for and base.device_type in except_for:
```

Reply:
Looks good to me, thank you for the update. Do you know which of the two is faster/more readable: `x != []`(works for all types, but would returns false if x is tuple) vs `len(x) == 0` (would return true for empty string, array dict and tuple would raise an exception if None or integer are passed as argument)?

Reply:
Could not use lambda. flake8 doesn't like binding it to name.

```
torch/testing/_internal/common_device_type.py:483:5: E731 do not assign a lambda expression, use a def
```

Reply:
@malfet You were right. Using just `if except_for and only_for` is best as it is most versatile irrespective of type.

Thanks!